src/prover.rs

```diff
@@ -75,6 +77,150 @@ pub struct Proof<F: IsField, CS: IsCommitmentScheme<F>> {
     pub w_zeta_omega_1: CS::Commitment,
 }
 
+impl<F, CS> Serializable for Proof<F, CS>
+where
+    F: IsField,
+    CS: IsCommitmentScheme<F>,
+    FieldElement<F>: ByteConversion,
+    CS::Commitment: Serializable,
+{
+    fn serialize(&self) -> Vec<u8> {
+        let field_elements = vec![
+            &self.a_zeta,
+            &self.b_zeta,
+            &self.c_zeta,
+            &self.s1_zeta,
+            &self.s2_zeta,
+            &self.z_zeta_omega,
+            &self.p_non_constant_zeta,
+            &self.t_zeta,
+        ];
+        let commitments = vec![
+            &self.a_1,
+            &self.b_1,
+            &self.c_1,
+            &self.z_1,
+            &self.t_lo_1,
+            &self.t_mid_1,
+            &self.t_hi_1,
+            &self.w_zeta_1,
+            &self.w_zeta_omega_1,
+        ];
+
+        let mut serialized_proof: Vec<u8> = Vec::new();
+
+        field_elements.iter().for_each(|element| {
+            let serialized_element = element.to_bytes_be();
+            serialized_proof.extend_from_slice(&serialized_element.len().to_be_bytes());
+            serialized_proof.extend_from_slice(&serialized_element);
+        });
+
+        commitments.iter().for_each(|commitment| {
+            let serialized_commitment = commitment.serialize();
+            serialized_proof.extend_from_slice(&serialized_commitment.len().to_be_bytes());
+            serialized_proof.extend_from_slice(&serialized_commitment);
+        });
+
+        serialized_proof
+    }
+}
+
+fn deserialize_field_element<F>(
+    bytes: &[u8],
+    separator_size: usize,
+    offset: usize,
+) -> Result<(usize, FieldElement<F>), DeserializationError>
+where
+    F: IsField,
+    FieldElement<F>: ByteConversion,
+{
+    let mut offset = offset;
+    let element_size_bytes: [u8; size_of::<usize>()] = bytes[offset..offset + separator_size]
+        .try_into()
+        .map_err(|_| DeserializationError::InvalidAmountOfBytes)?;
+    let element_size = usize::from_be_bytes(element_size_bytes);
+    offset += separator_size;
+    let field_element = FieldElement::from_bytes_be(&bytes[offset..offset + element_size])?;
+    offset += element_size;
+    Ok((offset, field_element))
+}
+
+fn deserialize_commitment<F, CS>(
+    bytes: &[u8],
+    separator_size: usize,
+    offset: usize,
+) -> Result<(usize, CS::Commitment), DeserializationError>
+where
+    F: IsField,
+    CS: IsCommitmentScheme<F>,
+    CS::Commitment: Deserializable,
+{
+    let mut offset = offset;
+    let element_size_bytes: [u8; size_of::<usize>()] = bytes[offset..offset + separator_size]
+        .try_into()
+        .map_err(|_| DeserializationError::InvalidAmountOfBytes)?;
+    let element_size = usize::from_be_bytes(element_size_bytes);
+    offset += separator_size;
+    let commitment = CS::Commitment::deserialize(&bytes[offset..offset + element_size])?;
```

Comment:
This may panic if I send a number of bytes between offset and element_size, use get